### PR TITLE
Add Flag To Prevent Unneccessary D->H Copies

### PR DESCRIPTION
**Pull request by njwhite**
By default Numba copies device arrays it creates back over host `np.array`s passed as kernel arguments once the kernel is finished. This is inefficient if the kernel only reads the arrays! This PR adds a flag to the `@jit` decorator to disable these post-call transfers for all raw `np.array` arguments for a kernel.

### CODE REVIEW

**Comment by sklam** [comment]
You know you can do `cuda_kernel[...](cuda.to_device(npy_ary))` to bypass the D->H.

**Comment by seibert** [comment]
I think I'd prefer to support this in the way that PyCUDA does, with their `In()`, `Out()` and `InOut()` wrappers:

https://documen.tician.de/pycuda/tutorial.html#shortcuts-for-explicit-memory-copies

In the future where we're all using CUDA managed memory, these functions can turn into prefetch hints instead of actual copies.

**Comment by njwhite** [comment]
@sklam yes, my code was getting rather verbose though (especially when passing in the stream). @seibert will try doing it like that, but as I have lots of parameters that are just In I might add a default to the config for untagged params?

**Comment by sklam** [comment]
@njwhite 

I agree that the PyCUDA style to mark individual arg is better than adding another flag into the `cuda.jit`.   But, if you want to reduce the verbosity of your source, you can consider adding a custom wrapper for the cuda kernels in your project.  The wrapper can apply `auto_device` on all the arguments but never copyback.

It may look something like:

```python
@nocopy
@cuda.jit
def cu_work(...):
   ...
```

Also, if your code can skip the copyback a lot of the time, you probably want to do the `cuda.to_device` very early on to get the numpy arrays into the GPU.

**Comment by sklam** [comment]
@njwhite, sorry for the slow reply.  I made a little patch that makes the argument preparation for a cuda kernel extensible.  It's in https://gist.github.com/sklam/a627270f9ecce13d88abb0dc1d5af7ad .
Also, there's an example script in it.  You can use it to implement the `nocopy` logic so you don't need the `CUDA_DEFAULT_ARGHINT` global config.

**Comment by njwhite** [comment]
Hi @sklam - I've added code based on your gist to the PR (and updated the tests!). I found it easier to make the extensions just map arguments, rather than intercepting them, so I didn't have to copy+paste the logic in `_prepare_args` into the decorator when I wanted to reuse it.

**Comment by njwhite** [comment]
@sklam ok, tests pass now. Can you re-review please?

File: numba/cuda/compiler.py [hunk 1]
```diff
@@ -589,16 +595,25 @@ def _prepare_args(self, ty, val, stream, retr, kernelargs):
         """
         Convert arguments to ctypes and append to kernelargs
         """
+
+        # map the arguments using any extension you've registered
+        ty, val = six.moves.reduce(
```
Comment (sklam) [inline numba/cuda/compiler.py]: Why use `reduce`?  All extensions will be applied from left to right.  Given the `nocopy` decorator is appending new extensions, I think extensions should be applied right-to-left.  Also, can you document what extension should do if it should fallthrough

**Comment by njwhite** [comment]
@sklam thanks for the review - I've pushed a commit to make the changes you suggested.

**Comment by stuartarchibald** [comment]
@sklam should this have a run through the build farm before merge?

**Comment by stuartarchibald** [comment]
@njwhite please could you merge the Numba master branch into the branch this PR is based on? There are a few critical fixes on master that need picking up to allow a smooth test of this patch on the build farm CUDA test builders. Thanks.

**Comment by njwhite** [comment]
@stuartarchibald done!

**Comment by stuartarchibald** [comment]
@njwhite :) thanks, will push it though the farm now.

**Comment by stuartarchibald** [comment]
Build farm is happy. Thanks for the extra fix @sklam. Merging.

**Pull request merged**